### ISSUE TITLE
Support for Linux Crypto API

### ISSUE THREAD

**Issue by ricsirigu** [hashcat/hashcat#1730]
Hi everyone,

we are trying to add a new custom cracking algorithm to bruteforce a disk encrypted using the crypto apis on linux kernel 2.4 
The algorithms we are going to use is Twofish with a key of 256 bits and the password is hashed with ripemd160.
I saw that Twofish and ripemd160 are already implemented in hashcat, 

Could someone give us some pointers on where to put our hands in order to implement this new algorithm? 

Thank you

**Comment by jsteube** [comment]
Which disk encryption are you talking about? Is that somehow different than the LUKS implementation we already have?

**Comment by ricsirigu** [comment]
@jsteube yes, it's different. The encryption scheme used is the one before LUKS, using the cryptoapi of the 2.4 kernel.

The old disk can be mounted with the current kernel using the command 

``` cryptsetup create --cipher twofish-cbc-plain --key-size 256 --hash ripemd160 crypted disk.raw```

in our case we know that the encrypted data contains an ext3 file system and this information is useful to know if the tried password is a correct guess or not.

Decrypting 16 bytes at offset 600h in a ext2/ext3 file system should produce 16 bytes of 00h.

Unlike LUKS, that has an header with the information of the cipher used and the keys encrypted, in the 2.4 kernel there is no header but only raw data.

**Comment by matrix** [comment]
Hi @ricsirigu, it's a known plaintext attack, interesting :)
I have little free time, but I try as soon as possible :)

**Comment by matrix** [comment]
This is KPA with AES256 and RIPEMD160 :)

```
$cryptoapi$9$2$d1d20e91a8f2e18881dc79369d8af761:hashcat
                                                 
Session..........: hashcat
Status...........: Cracked
Hash.Name........: Linux Crypto API
Hash.Target......: $cryptoapi$9$2$d1d20e91a8f2e18881dc79369d8af761
Time.Started.....: Sat Jan 23 23:30:57 2021 (0 secs)
Time.Estimated...: Sat Jan 23 23:30:57 2021 (0 secs)
Guess.Base.......: File (14500.pass)
Guess.Queue......: 1/1 (100.00%)
Speed.#1.........:     6024 H/s (0.01ms) @ Accel:1024 Loops:1 Thr:1 Vec:4
Recovered........: 1/1 (100.00%) Digests
Progress.........: 1/1 (100.00%)
Rejected.........: 0/1 (0.00%)
Restore.Point....: 0/1 (0.00%)
Restore.Sub.#1...: Salt:0 Amplifier:0-1 Iteration:0-1
Candidates.#1....: hashcat -> hashcat
```

**Comment by matrix** [comment]
@ricsirigu
this is KPA with Twofish256 and RIPEMD160

```
$cryptoapi$11$2$75f61d25151b674c00aca1ab68153698:hashcat
                                                 
Session..........: hashcat
Status...........: Cracked
Hash.Name........: Linux Crypto API
Hash.Target......: $cryptoapi$11$2$75f61d25151b674c00aca1ab68153698
Time.Started.....: Sun Jan 24 08:41:37 2021 (0 secs)
Time.Estimated...: Sun Jan 24 08:41:37 2021 (0 secs)
Guess.Base.......: File (14500.pass)
Guess.Queue......: 1/1 (100.00%)
Speed.#1.........:     7353 H/s (0.02ms) @ Accel:1024 Loops:1 Thr:1 Vec:4
Recovered........: 1/1 (100.00%) Digests
Progress.........: 1/1 (100.00%)
Rejected.........: 0/1 (0.00%)
Restore.Point....: 0/1 (0.00%)
Restore.Sub.#1...: Salt:0 Amplifier:0-1 Iteration:0-1
Candidates.#1....: hashcat -> hashcat

```

**Comment by ricsirigu** [comment]
> @ricsirigu
> this is KPA with Twofish256 and RIPEMD160
> 
> ```
> $cryptoapi$11$2$75f61d25151b674c00aca1ab68153698:hashcat
>                                                  
> Session..........: hashcat
> Status...........: Cracked
> Hash.Name........: Linux Crypto API
> Hash.Target......: $cryptoapi$11$2$75f61d25151b674c00aca1ab68153698
> Time.Started.....: Sun Jan 24 08:41:37 2021 (0 secs)
> Time.Estimated...: Sun Jan 24 08:41:37 2021 (0 secs)
> Guess.Base.......: File (14500.pass)
> Guess.Queue......: 1/1 (100.00%)
> Speed.#1.........:     7353 H/s (0.02ms) @ Accel:1024 Loops:1 Thr:1 Vec:4
> Recovered........: 1/1 (100.00%) Digests
> Progress.........: 1/1 (100.00%)
> Rejected.........: 0/1 (0.00%)
> Restore.Point....: 0/1 (0.00%)
> Restore.Sub.#1...: Salt:0 Amplifier:0-1 Iteration:0-1
> Candidates.#1....: hashcat -> hashcat
> ```

Hi @matrix! It looks awesome, how do you specify the plaintext?

**Comment by matrix** [comment]
The ciphertext is the one in bold below, and must be extracted from the target disk/image.

$cryptoapi$11$2$**75f61d25151b674c00aca1ab68153698**

For now the hash format is the following:
- Signature: _cryptoapi_
- Cipher/Hash type: _11_ (ripemd160, twofish)
- Cipher key size: _2_ (256, but also 128 and 192 are supported)
- Ciphertext: _75f61d25151b674c00aca1ab68153698_ (extracted from the encrypted image/disk)

@ricsirigu , can I ask you a question? :)
Why are you interested in this type of attack?
It's been about 16/17 years since the weakness of CryptoAPI on the 2.4 kernel was made public.

Thanks

**Comment by ricsirigu** [comment]
> The ciphertext is the one in bold below, and must be extracted from the target disk/image.
> 
> $cryptoapi$11$2$**75f61d25151b674c00aca1ab68153698**
> 
> For now the hash format is the following:
> 
> * Signature: _cryptoapi_
> * Cipher/Hash type: _11_ (ripemd160, twofish)
> * Cipher key size: _2_ (256, but also 128 and 192 are supported)
> * Ciphertext: _75f61d25151b674c00aca1ab68153698_ (extracted from the encrypted image/disk)
> 
> @ricsirigu , can I ask you a question? :)
> Why are you interested in this type of attack?
> It's been about 16/17 years since the weakness of CryptoAPI on the 2.4 kernel was made public.
> 
> Thanks

Sure, @matrix. I lost the password of an old hard drive encrypted with an old version of the Linux CryptoAPI and I was trying to recover its content.
If I understood correctly your code tries to bruteforce the ciphertext, right? 
Because I don't see a plaintext to perform a KPA attack or am I missing something?

**Comment by matrix** [comment]
> If I understood correctly your code tries to bruteforce the ciphertext, right?
> Because I don't see a plaintext to perform a KPA attack or am I missing something?

You missed something, but (in layman's terms) you'll be able to bruteforce the passphrase and, if you're lucky (so you find it), decrypt your drive.

**Comment by jsteube** [comment]
Support was added to hashcat GitHub master and will be part of the next release version.
